programs/mango-v4/src/instructions/openbook_v2_place_order.rs

```diff
@@ -0,0 +1,565 @@
+        // Double-check that we got the right account
+        let receiver_bank2 = receiver_bank_ai.load::<Bank>()?;
+        assert_eq!(receiver_bank2.group, group_key);
+        assert_eq!(receiver_bank2.token_index, receiver_token_index);
+    }
+
+    drop(retriever);
+
+    // No version check required, bank writable from v1
+
+    //
+    // Before-order tracking
+    //
+
+    let before_vault = ctx.accounts.vault.amount;
+
+    let before_oo_free_slots;
+    let before_had_bids;
+    let before_had_asks;
+    let before_oo = {
+        let open_orders = ctx.accounts.open_orders.load()?;
+        before_oo_free_slots = MAX_OPEN_ORDERS
+            - open_orders
+                .all_orders_in_use()
+                .collect::<Vec<&OpenOrder>>()
+                .len();
+        before_had_bids = open_orders.position.bids_base_lots != 0;
+        before_had_asks = open_orders.position.asks_base_lots != 0;
+        OpenOrdersSlim::from_oo_v2(&open_orders)
+    };
+
+    // Provide a readable error message in case the vault doesn't have enough tokens
+    let base_lot_size;
+    let quote_lot_size;
+    {
+        let openbook_market_external = ctx.accounts.openbook_v2_market_external.load()?;
+        base_lot_size = openbook_market_external.base_lot_size;
+        quote_lot_size = openbook_market_external.quote_lot_size;
+
+        // todo-pan: why i64? hope the cast doesnt fuck anything up
+        let needed_amount = match order.side {
+            OpenbookV2Side::Ask => (order.max_base_lots as u64 * base_lot_size as u64)
+                .saturating_sub(before_oo.native_base_free()),
+            OpenbookV2Side::Bid => (order.max_quote_lots_including_fees as u64)
+                .saturating_sub(before_oo.native_quote_free()),
+        };
+        if before_vault < needed_amount {
+            return err!(MangoError::InsufficentBankVaultFunds).with_context(|| {
+                format!(
+                    "bank vault does not have enough tokens, need {} but have {}",
+                    needed_amount, before_vault
+                )
+            });
+        }
+    }
+
+    //
+    // CPI to place order
+    //
+    let account_seeds = mango_account_seeds!(account.fixed);
+    cpi_place_order(
+        ctx.accounts,
+        &[account_seeds],
+        &order,
+        limit.try_into().unwrap(),
+    )?;
+
+    //
+    // After-order tracking
+    //
+    let open_orders = ctx.accounts.open_orders.load()?;
+    let after_oo_free_slots;
+    let after_oo = {
+        after_oo_free_slots = MAX_OPEN_ORDERS
+            - open_orders
+                .all_orders_in_use()
+                .collect::<Vec<&OpenOrder>>()
+                .len();
+        OpenOrdersSlim::from_oo_v2(&open_orders)
+    };
+    let oo_difference = OODifference::new(&before_oo, &after_oo);
+
+    //
+    // Track the highest bid and lowest ask, to be able to evaluate worst-case health even
+    // when they cross the oracle
+    //
```

Comment:
Is updating a min/max approximation even needed for openbook v2? Openbook v2 tracks the `locked_price` for each OpenOrder, which should be good enough. And since there can be only 24 open orders at a time, looping over them wouldn't be too expensive.

Reply:
That means we might be able to update max_quote and max_base to specific values based on the open orders state, instead of overestimating them!